### DISCUSSION TITLE
Adding a final image position & Ray tracing from point sources

### DISCUSSION THREAD

**Discussion by yunapark39** [mjhoptics/ray-optics#121] [Q&A]
Hello, 

Thank you very much for distributing the package!

I want to use a ZMX file from Thorlabs to make an optical path, and trace the rays from a list of point sources. 

I was not sure how to define the final imaging position after adding the zmx lens. From what I see, if an arbitrary surface were added, that would be considered an air, and trace_raw should trace to the final surface. 

However, I think I am not adding the surface correctly, because display and tracing would generate errors after adding a surface. What is the right way to change the optical paths between an imported lens? 
```
root_pth = Path(rayoptics.__file__).resolve().parent
opm, info = open_model(root_pth/'zemax/tests/ACA254-030-B-Zemax(ZMX).zmx', info=True)
sm  = opm['seq_model']
sm.gaps[0].thi=30
sm.add_surface([0,30])
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
                        do_draw_rays=True, do_paraxial_layout=False).plot()
theta=0.5
phi=0.0
pt0 = np.array([0.5, 1., 0.])
dir0 = np.array([np.sin(theta)*np.sin(phi),np.sin(theta)*np.sin(phi), np.cos(theta)]) 
path = sm.path(wl=855, start=1)
marg_ray = rt.trace_raw(path, pt0, dir0, wvl)
```

Additionally, I would like to trace rays from a list of point sources and output the final positions of the rays.  I was using trace_raw to trace each rays individually and the script seems to hang or take a long time. Could you advise how to approach this? Below is some of my scripts: 

```
def rand_rays(n_rays=10,NA=[1,1]):
    #ray angles are randomly generated angles over diameter of NA
    theta=NA[0]*np.random.rand(n_rays)
    phi=NA[1]*np.random.rand(n_rays)
    dirs = np.array([[np.sin(th)*np.sin(ph),np.sin(th)*np.sin(ph), np.cos(th)] for th in theta for ph in phi])
    return dirs    
def trace_rays(x,y,raydir,sm=None,rt=None):
    wavelength=780.0
    path = sm.path(wl=wavelength, start=1)
    pt0=np.array([x,y,0])
    dir0=raydir
    marg_ray = rt.trace_raw(path, pt0, dir0, wavelength)
  #  print(list_ray(marg_ray))
    posf=marg_ray[0][-1][0]
    return posf
def trace_from_coords(x,y,n_rays=10,NA=[1,1],sm=None,rt=None):
    raydirlist = rand_rays(n_rays,NA=NA)
    poslist=[]
    xlistf=[]
    ylistf=[]
    for raydir in raydirlist:
        posf=trace_rays(x,y,raydir,sm=sm,rt=rt)
        xlistf.append(posf[0])
        ylistf.append(posf[1])
    
    df0=pd.DataFrame({'x':x,'y':y})    
    df1=pd.DataFrame({'x':xlistf,'y':ylistf})    
    return df0,df1
if __name__ == '__main__':
    somex,somey=some_function()
    inputxy = zip(somex,somey,repeat(n_rays),repeat(NA),repeat(sm),repeat(rt))
    with Pool(num_cores) as pool:
        results=list(pool.starmap(trace_from_coords, tqdm(inputxy, total=len(gex)**2)))    
```

Thank you !

**Comment by mjhoptics** [comment]
Hello @yunapark39,
I looked at the first part of your question and put together a Jupyter notebook that might be helpful. 
[issue121_investigation.pdf](https://github.com/mjhoptics/ray-optics/files/11439112/issue121_investigation.pdf)
There are some functions in the `analyses` module that can be used to trace lists of rays. We can discuss when the model is clear.
Mike Hayford

**Comment by mjhoptics** [comment]
## first look, issue #121

[Issue 121 GitHub discussion](https://github.com/mjhoptics/ray-optics/discussions/121#discussion-5178785)

- Restore a ThorLabs doublet
- flip it around
- set up object points in the focal plane.


```python
%matplotlib inline
```


```python
from rayoptics.environment import *

isdark = False
```

## Open the ZMX file, in the same directory as this notebook


```python
opm, info = open_model(Path.cwd() / "ACA254-030-B-Zemax(ZMX).zmx", info=True)
```

## Review the `info` from reading the ZMX file

Nothing exceptional to note for this file


```python
info
```




    ({'VERS': '150514 1707 30221',
      'pupil': ('aperture', 'object', 'epd'),
      'FTYP': '0 0 1 3 0 0 0',
      'STANDARD': 7,
      '# clear ap': 7,
      'conj type': 'infinite',
      '# surfs': 7,
      '# wvls': 5,
      'fov': ('field', 'object', 'angle'),
      '# fields': 1,
      'GCAT': ['SCHOTT', 'CDGM', 'THORSHOP'],
      'glass found': 2,
      'encoding': 'utf-16'},
     {})



## Set up the standard set of aliases for **ray-optics**


```python
sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
```

## Examine the ThorLabs model

### list the sequential model, optical specs, and first order data

It is set up as an object at infinity, an epd of 20 -> f/1.5 aperture, and a wavelength range in the near IR.


```python
sm.list_model()
```

                  c            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:     0.000000      3.00000       air             1      10.000
     Stop:     0.050055      7.79328    N-SSK2             1      11.125
        3:    -0.037187      1.24393       air             1      11.125
        4:    -0.049539      9.26279    N-SF57             1      10.125
        5:    -0.026803      20.6784       air             1      11.125
      Img:     0.000000      0.00000                       1    0.055039



```python
listobj(osp)
```

    aperture: object epd; value=20.0
    field: object angle; value=1.0
    x,y=0.0 vlx= 0.000 vux= 0.000 vly= 0.000 vuy= 0.000
    central wavelength= 1015.0000 nm
    wavelength (weight) =  706.5000 (1.000),  855.0000 (1.000), 1015.0000 (1.000)*,  925.0000 (1.000),  750.0000 (1.000)
    focus shift=0.0
    



```python
pm.first_order_data()
```

    efl               30.11
    ffl              -23.33
    pp1                6.78
    bfl               21.02
    ppk               9.082
    m            -3.011e-09
    red          -3.322e+08
    obj_dist          1e+10
    obj_ang               1
    enp_dist              3
    enp_radius           10
    na obj            1e-09
    n obj                 1
    img_dist          21.02
    img_ht           0.5255
    exp_dist         -13.06
    exp_radius        11.44
    f/# img           1.505
    na img          -0.3152
    n img                 1
    optical invariant       0.1746
    



```python
pm.list_model()
```

               ax_ht        pr_ht       ax_slp       pr_slp         power          tau        index    type
     0:            0 -1.74551e+08        1e-09    0.0174551             0        1e+10      1.00000    dummy
     1:           10   -0.0523652        1e-09    0.0174551             0            3      1.00000    transmit
     2:           10  1.09544e-08    -0.305094    0.0174551    0.03050943        4.842      1.60952    transmit
     3:      8.52273    0.0845174    -0.498273    0.0155394    0.02266626      1.24393      1.00000    transmit
     4:      7.90292     0.103847    -0.180012    0.0197214   -0.04027128       5.1093      1.81293    transmit
     5:      6.98318      0.20461    -0.332165    0.0152633    0.02178851      20.6784      1.00000    transmit
     6:      0.11452     0.520231    -0.332165    0.0152633             0            0      1.00000    dummy


## Lens layout of the model with no changes looks OK


```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
                        do_draw_rays=True, do_paraxial_layout=False).plot()
```


    
![output_15_0](https://github.com/mjhoptics/ray-optics/assets/33294031/93343224-b7c0-4e47-851f-340e1468170b)
    


## Model changes

It looked like you wanted to do the ray tracing from the short conjugate to long. You can do this by flipping the lens assembly around using the [flip](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.optical.opticalmodel.html#rayoptics.optical.opticalmodel.OpticalModel.flip) command. For this case we'll flip from surface 1 to 5, i.e. the doublet and diaphragm, but not the object and image surfaces.


```python
opm.flip(1,5)
```


```python
sm.list_model()
```

                  c            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:     0.026803      9.26279    N-SF57             1      11.125
        2:     0.049539      1.24393       air             1      10.125
        3:     0.037187      7.79328    N-SSK2             1      11.125
     Stop:    -0.050055      3.00000       air             1      11.125
        5:    -0.000000      20.6784       air             1      10.000
      Img:     0.000000      0.00000                       1    0.055039


## Update object and image distances

The flip operation didn't affect the object and image distances. We'll copy the original image distance to the object and set the new image distance to 100 (because I'm not sure of your intent here...). 


```python
sm.gaps[0].thi = sm.gaps[-1].thi
sm.gaps[-1].thi = 100.
```

## Update the aperture and field specifications

The original Thorlabs model used an object angle specification for the field. Since the object distance is finite now, an object height specification can be used. I've arbitrarily set the max field height to 2.5


```python
osp['fov'].key = 'field', 'object', 'height'
osp['fov'].value = 2.5
osp['fov'].set_from_list([0., 2.5])
```

The aperture specification in terms of entrance pupil diameter can be retained in the flipped configuration. However, the exit pupil diameter of the original configuration should be used as the new entrance pupil diameter.

Finally, to get the layout to show the object, manually set the conjugate_type to 'finite'. I think this is a bug, not a feature...


```python
osp['pupil'].value = 22.88
opm['specsheet'].conjugate_type = 'finite'
```

## Set up the system wavelength(s)

The Zemax model had a 5 wavelength spectral sampling. Here's a sampling using the wavelengths in your example.


```python
osp['wvls'] = WvlSpec([(780., 1), (855., 1.)], ref_wl=1)
```

## After making the changes, do update_model()


```python
opm.update_model()
```

## Lens layout shows reasonable flipped model


```python
layout_plt1 = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
                        do_draw_rays=True, do_paraxial_layout=False).plot()
```


    
![output_30_0](https://github.com/mjhoptics/ray-optics/assets/33294031/36b73ad9-5bb1-481d-8ee4-85393e94f1f2)
    



```python
sm.list_model()
```

                  c            t        medium     mode   zdr      sd
      Obj:     0.000000      20.6784       air             1      1.0000
        1:     0.026803      9.26279    N-SF57             1      11.125
        2:     0.049539      1.24393       air             1      10.125
        3:     0.037187      7.79328    N-SSK2             1      11.125
     Stop:    -0.050055      3.00000       air             1      11.125
        5:    -0.000000      100.000       air             1      10.000
      Img:     0.000000      0.00000                       1    0.055039



```python
listobj(osp)
```

    aperture: object epd; value=22.88
    field: object height; value=2.5
    x,y=0.0 vlx= 0.000 vux= 0.000 vly= 0.000 vuy= 0.000
    y=2.5 vly= 0.000 vuy= 0.000 vlx= 0.000 vux= 0.000
    central wavelength=  855.0000 nm
    wavelength (weight) =  780.0000 (1.000),  855.0000 (1.000)*
    focus shift=0.0
    



```python
pm.first_order_data()
```

    efl                  30
    ffl              -20.94
    pp1               9.061
    bfl               23.22
    ppk               6.776
    m                 115.1
    red            0.008687
    obj_dist          20.68
    obj_ang          -4.198
    enp_dist          13.38
    enp_radius        11.44
    na obj           0.3184
    n obj                 1
    img_dist          -3430
    img_ht            287.8
    exp_dist          -3533
    exp_radius           10
    f/# img          -171.4
    na img         0.002918
    n img                 1
    optical invariant      -0.8397
    



```python
pm.list_model()
```

               ax_ht        pr_ht       ax_slp       pr_slp         power          tau        index    type
     0:            0          2.5     0.335881   -0.0734005             0      20.6784      1.00000    dummy
     1:      6.94548     0.982193     0.183219   -0.0949891    0.02197998      5.08925      1.82007    transmit
     2:      7.87793      0.49877     0.503261   -0.0747264   -0.04062518      1.24393      1.00000    transmit
     3:      8.50395     0.405815     0.309541   -0.0839709    0.02278006      4.83281      1.61258    transmit
     4:      9.99991  2.22045e-16   0.00291768   -0.0839709     0.0306626            3      1.00000    transmit
     5:      10.0087    -0.251913   0.00291768   -0.0839709            -0          100      1.00000    transmit
     6:      10.3004     -8.64901   0.00291768   -0.0839709             0            0      1.00000    dummy


## Further analysis

The raytr.analysis module has classes and functions for defining and tracing lists or grids of rays.


```python

```

**Reply by yunapark39** [reply]
Hi, 

Thanks for the prompt response! 

I don't think I needed to flip my model, but I was able to correct the spacing by changing the gap thickness as you did..

I tried using trace_list_of_rays, but the computation time is still very long. Is trace_list_of_rays the best method or is there some option to make things faster such as ignoring the rays that go out of the range (or is the pupil function doing this already)? 

Also, have you run the trace_list_of_rays function in parallel computation? I tried running the following script in parallel, but windows needs each of the defined variables like opm to be passed to the function, and I couldn't figure out a good way. My script was: 
```

def trace_from_coords(x,y,n_rays=10,NA=[1,1],wavelength=780,opm=None):
    rays=make_rays(x,y,n_rays,NA,wavelength)
    rays_out=ras.trace_list_of_rays(opm,rays)
    poslist=[x[0][0][0] for x in rays_out]    
    xlistf=[x[0] for x in poslist]
    ylistf=[x[1] for x in poslist]
    xlist=np.ones_like(xlistf)*x
    ylist=np.ones_like(xlistf)*y
    df0=pd.DataFrame({'x':xlist,'y':ylist})    
    df1=pd.DataFrame({'x':xlistf,'y':ylistf})    
   # print('number of rays initial:%d final:%d'%(len(xlist),len(xlistf)))
    return df0,df1
from itertools import repeat
from multiprocessing import Pool, cpu_count

inputxy = zip(gex,gey,repeat(n_rays),repeat(NA),repeat(wavelength),repeat(opm))
with Pool(num_cores) as pool:
       results=list(pool.starmap(trace_from_coords, tqdm(inputxy,total=len(gex)**2)))    
        
```

**Comment by mjhoptics** [comment]
I've spent very little time looking at performance, the primary concern has been correctness. I haven't spent any time looking at parallel performance. My previous experience with the Dev team multi-threading *LightTools* was 1) the first attempt running something in parallel runs slower than running it single-threaded and 2) you need to measure everything. The `multiprocess` module looks like it spins up separate processes and that would require a lot of work to amortize.

The one area I've seen dramatic speedups in python is with numpy array arithmetic. If you set your code up right, you can pass an array or scalar into a function and have the array vector processed automatically. I've seen big speedups vs. for loops. I think trying to use numpy array calculations as much as possible in the ray trace to trace *arrays of rays* has the biggest potential for performance improvement. 

There are a few things I've been curious about the ray trace that could have a performance impact. I'll take a look and if I find anything significant, I'll report on it.

Mike
